Source/Turbo Navigator/TurboNavigationHierarchyController.swift

```diff
@@ -0,0 +1,203 @@
+            return topViewController.isMember(of: type(of: controller))
+        }
+        return false
+    }
+
+    private func visitingPreviousPage(on navigationController: UINavigationController, with controller: UIViewController, via url: URL) -> Bool {
+        guard navigationController.viewControllers.count >= 2 else { return false }
+
+        let previousController = navigationController.viewControllers[navigationController.viewControllers.count - 2]
+        if let previousVisitable = previousController as? VisitableViewController {
+            return previousVisitable.visitableURL == url
+        }
+        return type(of: previousController) == type(of: controller)
+    }
+
+    private func pop(via proposal: VisitProposal) {
+        if navigationController.presentedViewController != nil {
+            if modalNavigationController.viewControllers.count == 1 {
+                navigationController.dismiss(animated: proposal.animated)
+            } else {
+                modalNavigationController.popViewController(animated: proposal.animated)
+            }
+        } else {
+            navigationController.popViewController(animated: proposal.animated)
+        }
+    }
+
+    private func replace(with controller: UIViewController, via proposal: VisitProposal) {
+        switch proposal.context {
+        case .default:
+            navigationController.dismiss(animated: proposal.animated)
+            navigationController.replaceLastViewController(with: controller)
+            if let visitable = controller as? Visitable {
+                delegate.visit(visitable, on: .main, with: proposal.options)
+            }
+        case .modal:
+            if navigationController.presentedViewController != nil {
+                modalNavigationController.replaceLastViewController(with: controller)
+            } else {
+                modalNavigationController.setViewControllers([controller], animated: false)
+                modalNavigationController.setModalPresentationStyle(via: proposal)
+                navigationController.present(modalNavigationController, animated: proposal.animated)
+            }
+            if let visitable = controller as? Visitable {
+                delegate.visit(visitable, on: .modal, with: proposal.options)
+            }
+        }
+    }
+
+    private func refresh(via proposal: VisitProposal) {
+        if navigationController.presentedViewController != nil {
+            if modalNavigationController.viewControllers.count == 1 {
+                navigationController.dismiss(animated: proposal.animated)
+                refreshIfTopViewControllerIsVisitable(from: .main)
+            } else {
+                modalNavigationController.popViewController(animated: proposal.animated)
+                refreshIfTopViewControllerIsVisitable(from: .modal)
+            }
+        } else {
+            navigationController.popViewController(animated: proposal.animated)
```

Comment:
Guess this isn't required here when the `presentedViewController` is nil because we aren't displaying a modal window and we will refreshing the previous item in the navigation stack